segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3StorageTest.java

```diff
@@ -0,0 +1,265 @@
+                                                    .build();
+        if (client == null) {
+            createStorage();
+
+            try {
+                client.createBucket("kanpravegatest");
+            } catch (S3Exception e) {
+                if (!e.getErrorCode().equals("BucketAlreadyOwnedByYou")) {
+                    throw e;
+                }
+            }
+            List<ObjectKey> keys = client.listObjects("kanpravegatest").getObjects().stream().map((object) -> {
+                return new ObjectKey(object.getKey());
+            }).collect(Collectors.toList());
+
+            if (!keys.isEmpty()) {
+                client.deleteObjects(new DeleteObjectsRequest("kanpravegatest").withKeys(keys));
+            }
+        }
+    }
+
+    @After
+    public void tearDown() throws Exception {
+            client.shutdown();
+            client = null;
+            s3Proxy.stop();
+    }
+
+    @Override
+    protected Storage createStorage() {
+        S3Config s3Config = null;
+        URI uri = URI.create(endpoint);
+        s3Config = new S3Config(uri);
+        s3Config.withIdentity(adapterConfig.getAccessKey()).withSecretKey(adapterConfig.getSecretKey());
+
+        client = new S3JerseyClientWrapper(s3Config, aclMap);
+
+        ExtendedS3Storage storage = new ExtendedS3Storage(client, adapterConfig, executorService());
+        return storage;
+    }
+
+    @Override
+    protected SegmentHandle createHandle(String segmentName, boolean readOnly, long epoch) {
+        FileChannel channel = null;
+        if (readOnly) {
+            return ExtendedS3SegmentHandle.getReadHandle(segmentName);
+        } else {
+            return ExtendedS3SegmentHandle.getWriteHandle(segmentName);
+        }
+    }
+
+    /**
+     * Wrapper over S3JerseyClient. This implements ACLs, multipart copy and multiple writes to the same object on top of S3Proxy implementation.
+     */
+    private static class S3JerseyClientWrapper extends S3JerseyClient {
+        private final ConcurrentMap<String, AclSize> aclMap;
+
+        public S3JerseyClientWrapper(S3Config s3Config, ConcurrentMap<String, AclSize> aclMap) {
+            super(s3Config);
+            this.aclMap = aclMap;
+        }
+
+        @Override
+        public PutObjectResult putObject(PutObjectRequest request) {
+            S3ObjectMetadata metadata = request.getObjectMetadata();
+
+           if ( request.getObjectMetadata() != null ) {
+            request.setObjectMetadata(null);
+           }
+            PutObjectResult retVal = super.putObject(request);
+           if (request.getAcl() != null) {
+               long size = 0;
+               if (request.getRange() != null) {
+                   size = request.getRange().getLast() -1;
+               }
+               aclMap.put(request.getKey(), new AclSize(request.getAcl(),
+                       size));
+           }
+           return retVal;
+        }
+
+        @Synchronized
```

Comment:
Why is this putObject call synchronized but the one above is not? Is it intended that one using the single arg should be able to proceed in parallel to this one?